Provide a swift implementation for CommandEncoder::clearBuffer

nmahendru opened the pull request:
#### 6a9a052fc36c271ed249b88a35d652594a5107c7
<pre>
Provide a swift implementation for CommandEncoder::clearBuffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=280767">https://bugs.webkit.org/show_bug.cgi?id=280767</a>
<a href="https://rdar.apple.com/137134098">rdar://137134098</a>

Reviewed by Mike Wyrzykowski.

It&apos;s a 1 to 1 replacement of the Obj-C version.

Testing:
compile with make release BUILD_WEBKIT_OPTIONS=&quot;--webGpuSwift&quot;.
Then run:
1. http/tests/webgpu/webgpu/api/validation/encoding/cmds/clearBuffer.html
2. http/tests/webgpu/webgpu/api/operation/command_buffer/clearBuffer.html

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/CommandEncoder.h:
(retainCommandEncoder):
(releaseCommandEncoder):
(WebGPU::CommandEncoder::device const): Deleted.
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::generateInvalidEncoderStateError):
(WebGPU::CommandEncoder::clearBuffer):
* Source/WebGPU/WebGPU/CommandEncoder.swift: Copied from Source/WebGPU/WebGPU/CommandsMixin.h.
(clearBuffer(_:buffer:offset:size:)):
(WebGPU.clearBuffer(_:offset:size:)):
* Source/WebGPU/WebGPU/CommandsMixin.h:
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::generateAValidationError):
* Source/WebGPU/WebGPU/WebGPUExt.h:

Canonical link: <a href="https://commits.webkit.org/284713@main">https://commits.webkit.org/284713@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/439c6fac791cee722de20db311f13b7a5e77b5bf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70222 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49628 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22986 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74311 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21392 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72339 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57428 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21240 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55685 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14170 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73288 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45189 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60559 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36147 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41848 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17993 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19761 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/63343 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63776 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18338 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76030 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/69469 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14448 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17577 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63398 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14490 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60625 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63334 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15576 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11366 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4989 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/91250 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45430 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/197 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/19888 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46504 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47781 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46246 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -0,0 +1,66 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+private let wgpuWholeSize: UInt64 = 0xffffffffffffffff
```
No constant in Swift for UInt64.max? Just asking, I don't know

mwyrzykowski commented on Source/WebGPU/WebGPU/WebGPUExt.h:
```diff
@@ -168,6 +168,8 @@ WGPU_EXPORT void wgpuDeviceClearUncapturedErrorCallback(WGPUDevice device) WGPU_
 #include "Buffer.h"
 #include "Device.h"
 #include "Queue.h"
+#include "CommandsMixin.h"
```
minor but should these be sorted?

mwyrzykowski reviewed (approved):
Thank you!

nmahendru commented on Source/WebGPU/WebGPU/Queue.swift:
```diff
@@ -28,11 +28,11 @@ private let largeBufferSize = 32 * 1024 * 1024
     queue: WebGPU.Queue, buffer: WebGPU.Buffer, bufferOffset: UInt64, data: SpanUInt8
 ) {
-    queue.writeBuffer(mtlBuffer: buffer.buffer(), bufferOffset: bufferOffset, data: data)
+    queue.writeBuffer(buffer: buffer, bufferOffset: bufferOffset, data: data)
```
@rr-codes fixed the `any` -> `some` here.
Actually I just removed the MTLBuffer and just passed around the WebGPU.Buffer instead.

rr-codes commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -0,0 +1,67 @@
+            let (subtractionResult, didOverflow) = initialSize.subtractingReportingOverflow(offset)
+            if didOverflow {
+                self.device().generateAValidationError(
+                    Swift.String("CommandEncoder::clearBuffer(): offset > buffer.size"))
```
```suggestion
                    "CommandEncoder::clearBuffer(): offset > buffer.size")
```

does this not work?

nmahendru replied:
For some reason WTF.String has been imported here as String. So if I use the default then it becomes a WTF.String.

rr-codes replied:
even string literals in the Swift file get turned into WTF strings? I would not expect that, and we should definitely try to fix that if so

nmahendru replied:
fixed now. This is the culprit. 
https://github.com/WebKit/WebKit/blob/2509a2481fe05555bffdeb1afa21471afdc0fd7e/Source/WTF/wtf/text/WTFString.h#L559

Although, I think the swift compiler should stop that from happening.

rr-codes commented on Source/WebGPU/WebGPU/Queue.swift:
```diff
@@ -28,11 +28,11 @@ private let largeBufferSize = 32 * 1024 * 1024
 
 extension WebGPU.Queue {
-    public func writeBuffer(mtlBuffer: any MTLBuffer, bufferOffset: UInt64, data: SpanUInt8) {
+    public func writeBuffer(buffer: WebGPU.Buffer, bufferOffset: UInt64, data: SpanUInt8) {
```
```suggestion
    public func writeBuffer(_ buffer: WebGPU.Buffer, bufferOffset: UInt64, data: SpanUInt8) {
```

I think the `buffer` parameter name can be omitted since it's in the function name

rr-codes commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -0,0 +1,67 @@
+            return
+        }
+        if size == UInt64.max {
+            let initialSize: UInt64 = buffer.initialSize()
```
is the `: UInt64` needed?

nmahendru committed:
Provide a swift implementation for CommandEncoder::clearBuffer
https://bugs.webkit.org/show_bug.cgi?id=280767
rdar://137134098

Reviewed by Mike Wyrzykowski.

It's a 1 to 1 replacement of the Obj-C version.

Testing:
compile with make release BUILD_WEBKIT_OPTIONS="--webGpuSwift".
Then run:
1. http/tests/webgpu/webgpu/api/validation/encoding/cmds/clearBuffer.html
2. http/tests/webgpu/webgpu/api/operation/command_buffer/clearBuffer.html

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/CommandEncoder.h:
(retainCommandEncoder):
(releaseCommandEncoder):
(WebGPU::CommandEncoder::device const): Deleted.
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::generateInvalidEncoderStateError):
(WebGPU::CommandEncoder::clearBuffer):
* Source/WebGPU/WebGPU/CommandEncoder.swift: Copied from Source/WebGPU/WebGPU/CommandsMixin.h.
(clearBuffer(_:buffer:offset:size:)):
(WebGPU.clearBuffer(_:offset:size:)):
* Source/WebGPU/WebGPU/CommandsMixin.h:
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::generateAValidationError):
* Source/WebGPU/WebGPU/WebGPUExt.h:

Canonical link: https://commits.webkit.org/284713@main
